Use camunda-preferred connection pool (HikariCP)

Periodically, we see issues like this:

```
org.springframework.transaction.CannotCreateTransactionException: Could not open JDBC Connection for transaction; nested exception is java.sql.SQLNonTransientConnectionException: Could not connect to address=(host=cws-db-g71bt.cbtgh8rlzfls.us-gov-west-1.rds.amazonaws.com)(port=3306)(type=master) : Socket fail to connect to host:cws-db-g71bt.cbtgh8rlzfls.us-gov-west-1.rds.amazonaws.com, port:3306. Cannot assign requested address (connect failed)
at org.springframework.jdbc.datasource.DataSourceTransactionManager.doBegin(DataSourceTransactionManager.java:289)
at org.springframework.transaction.support.AbstractPlatformTransactionManager.getTransaction(AbstractPlatformTransactionManager.java:377)
at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:130)
at org.camunda.bpm.engine.spring.SpringTransactionInterceptor.execute(SpringTransactionInterceptor.java:44)
at org.camunda.bpm.engine.impl.interceptor.ProcessApplicationContextInterceptor.execute(ProcessApplicationContextInterceptor.java:70)
at org.camunda.bpm.engine.impl.interceptor.LogInterceptor.execute(LogInterceptor.java:33)
at org.camunda.bpm.engine.impl.RuntimeServiceImpl.setVariables(RuntimeServiceImpl.java:394)
at org.camunda.bpm.engine.impl.RuntimeServiceImpl.setVariables(RuntimeServiceImpl.java:384)
at org.camunda.bpm.engine.impl.RuntimeServiceImpl.setVariable(RuntimeServiceImpl.java:371)
at jpl.cws.engine.CwsExternalTaskThread.setStringVariable(CwsExternalTaskThread.java:453)
at jpl.cws.engine.CwsExternalTaskThread.setOutputVariable(CwsExternalTaskThread.java:461)
at jpl.cws.engine.CwsExternalTaskThread.executeTask(CwsExternalTaskThread.java:571)
at jpl.cws.engine.CwsExternalTaskThread.runTask(CwsExternalTaskThread.java:231)
at jpl.cws.engine.CwsExternalTaskThread.run(CwsExternalTaskThread.java:108)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
at java.lang.Thread.run(Thread.java:748)
Caused by: java.sql.SQLNonTransientConnectionException: Could not connect to address=(host=cws-db-g71bt.cbtgh8rlzfls.us-gov-west-1.rds.amazonaws.com)(port=3306)(type=master) : Socket fail to connect to host:cws-db-g71bt.cbtgh8rlzfls.us-gov-west-1.rds.amazonaws.com, port:3306. Cannot assign requested address (connect failed)
at org.mariadb.jdbc.internal.util.exceptions.ExceptionFactory.createException(ExceptionFactory.java:73)
at org.mariadb.jdbc.internal.util.exceptions.ExceptionFactory.create(ExceptionFactory.java:192)
at org.mariadb.jdbc.internal.protocol.AbstractConnectProtocol.connectWithoutProxy(AbstractConnectProtocol.java:1392)
at org.mariadb.jdbc.internal.util.Utils.retrieveProxy(Utils.java:635)
at org.mariadb.jdbc.MariaDbConnection.newConnection(MariaDbConnection.java:150)
at org.mariadb.jdbc.Driver.connect(Driver.java:89)
at org.springframework.jdbc.datasource.SimpleDriverDataSource.getConnectionFromDriver(SimpleDriverDataSource.java:140)
at org.springframework.jdbc.datasource.AbstractDriverBasedDataSource.getConnectionFromDriver(AbstractDriverBasedDataSource.java:196)
at org.springframework.jdbc.datasource.AbstractDriverBasedDataSource.getConnection(AbstractDriverBasedDataSource.java:159)
at org.springframework.jdbc.datasource.DataSourceTransactionManager.doBegin(DataSourceTransactionManager.java:246)
... 16 more
Caused by: java.sql.SQLNonTransientConnectionException: Socket fail to connect to host:cws-db-g71bt.cbtgh8rlzfls.us-gov-west-1.rds.amazonaws.com, port:3306. Cannot assign requested address (connect failed)
at org.mariadb.jdbc.internal.util.exceptions.ExceptionFactory.createException(ExceptionFactory.java:73)
at org.mariadb.jdbc.internal.util.exceptions.ExceptionFactory.create(ExceptionFactory.java:183)
at org.mariadb.jdbc.internal.protocol.AbstractConnectProtocol.createSocket(AbstractConnectProtocol.java:255)
at org.mariadb.jdbc.internal.protocol.AbstractConnectProtocol.createConnection(AbstractConnectProtocol.java:519)
at org.mariadb.jdbc.internal.protocol.AbstractConnectProtocol.connectWithoutProxy(AbstractConnectProtocol.java:1387)
... 23 more
Caused by: java.net.ConnectException: Cannot assign requested address (connect failed)
at java.net.PlainSocketImpl.socketConnect(Native Method)
at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
at java.net.Socket.connect(Socket.java:589)
at org.mariadb.jdbc.internal.protocol.AbstractConnectProtocol.createSocket(AbstractConnectProtocol.java:250)
... 25 more

```
I’m thinking this is the the camunda engine trying to progress the workflow itself..  I think CWS core should be upgraded to their preferred connection pool implementation.
see:  https://camunda.com/best-practices/_book/#_tuning_the_database_connection_pool

Camunda actually recommends we use this connection pool: https://github.com/brettwooldridge/HikariCP

--- a/create_server_dist.sh
+++ b/create_server_dist.sh
@@ -110,6 +110,7 @@ rm -rf ${CWS_TOMCAT_ROOT}/webapps/docs
 print 'Installing DB drivers to Tomcat...'
 cp ${ROOT}/cws-installer/cws-installer-libs/mysql-connector-java-*.jar ${TOMCAT_LIB_DIR}
 cp ${ROOT}/cws-installer/cws-installer-libs/mariadb-java-client-*.jar  ${TOMCAT_LIB_DIR}
+cp ${ROOT}/cws-installer/cws-installer-libs/HikariCP-*.jar             ${TOMCAT_LIB_DIR}
 
 print 'Installing core libraries to Tomcat...'
 cp ${ROOT}/cws-core/target/cws-core.jar                     ${TOMCAT_LIB_DIR}

--- a/cws-adaptation-engine/pom.xml
+++ b/cws-adaptation-engine/pom.xml
@@ -89,7 +89,13 @@
 			<groupId>mysql</groupId>
 			<artifactId>mysql-connector-java</artifactId>
 		</dependency>
-		
+
+		<!-- HikariCP database connection pool -->
+		<dependency>
+			<groupId>com.zaxxer</groupId>
+			<artifactId>HikariCP</artifactId>
+		</dependency>
+
 		<!-- Spring MVC framework -->
 		<dependency>
 			<groupId>org.springframework</groupId>

--- a/cws-engine-service/pom.xml
+++ b/cws-engine-service/pom.xml
@@ -191,6 +191,12 @@
 			<groupId>mysql</groupId>
 			<artifactId>mysql-connector-java</artifactId>
 		</dependency>
+
+		<!-- HikariCP database connection pool -->
+		<dependency>
+			<groupId>com.zaxxer</groupId>
+			<artifactId>HikariCP</artifactId>
+		</dependency>
 		
 		<!-- Spring MVC framework -->
 		<dependency>

--- a/cws-installer/pom.xml
+++ b/cws-installer/pom.xml
@@ -67,6 +67,12 @@
 			<groupId>mysql</groupId>
 			<artifactId>mysql-connector-java</artifactId>
 		</dependency>
+
+		<!-- HikariCP database connection pool -->
+		<dependency>
+			<groupId>com.zaxxer</groupId>
+			<artifactId>HikariCP</artifactId>
+		</dependency>
 		
 		<dependency>
 			<groupId>junit</groupId>

--- a/cws-installer/src/main/java/jpl/cws/task/CwsInstaller.java
+++ b/cws-installer/src/main/java/jpl/cws/task/CwsInstaller.java
@@ -2026,13 +2026,6 @@ public class CwsInstaller {
 		content = content.replace("__CWS_SHUTDOWN_PORT__",         cws_shutdown_port);
 		content = content.replace("__CWS_TOMCAT_CONF_DIR__",       cws_tomcat_conf);
 
-		if (cws_worker_type.equals("run_external_tasks_only")) {
-			content = content.replace("__CWS_DB_NUM_CONNECTIONS__", "2");
-		}
-		else {
-			content = content.replace("__CWS_DB_NUM_CONNECTIONS__", "1");
-		}
-
 		writeToFile(filePath, content);
 		copy(
 			Paths.get(config_work_dir + SEP + "tomcat_conf" + SEP + "server.xml"),

--- a/cws-service/pom.xml
+++ b/cws-service/pom.xml
@@ -126,6 +126,12 @@
 					<artifactId>mysql-connector-java</artifactId>
 				</dependency>
 
+				<!-- HikariCP database connection pool -->
+				<dependency>
+					<groupId>com.zaxxer</groupId>
+					<artifactId>HikariCP</artifactId>
+				</dependency>
+
 				<dependency>
 					<groupId>commons-fileupload</groupId>
 					<artifactId>commons-fileupload</artifactId>

--- a/install/tomcat_conf/server.xml
+++ b/install/tomcat_conf/server.xml
@@ -49,76 +49,37 @@
               factory="org.apache.catalina.users.MemoryUserDatabaseFactory"
               pathname="conf/tomcat-users.xml" />
 
-    <!-- Additional properties can be set to avoid a common exception
-         regarding connection timeout in the connection pool:
-          testOnBorrow="true"
-          validationQuery="SELECT 1"
-    -->
     <Resource name="jdbc/ProcessEngine"
               auth="Container"
               type="javax.sql.DataSource"
-              factory="org.apache.tomcat.jdbc.pool.DataSourceFactory"
-              uniqueResourceName="process-engine"
+              factory="com.zaxxer.hikari.HikariJNDIFactory"
+              poolName="camunda-pool"
+
+              maximumPoolSize="10"
+              connectionTimeout="300000"
+
               driverClassName="__CWS_DB_DRIVER__"
-              url="__CWS_DB_URL__"
+              jdbcUrl="__CWS_DB_URL__"
               username="__CWS_DB_USERNAME__"
               password="__CWS_DB_PASSWORD__"
-              
-              initialSize="__CWS_DB_NUM_CONNECTIONS__"
-              minIdle="__CWS_DB_NUM_CONNECTIONS__"
-              
-              maxIdle="20"
-              maxActive="20"
-              
-              validationQuery="SELECT 1"
-              validationInterval="30000"
-              logValidationErrors="true"
 
-              timeBetweenEvictionRunsMillis="30000"
-              minEvictableIdleTimeMillis="60000"
-              removeAbandoned="true"
-              removeAbandonedTimeout="900"
-              
-              testOnBorrow="true"
-              testWhileIdle="true"
-              testOnReturn="true"
-
-              logAbandoned="true"
-
-              defaultTransactionIsolation="READ_COMMITTED" />
+              transactionIsolation="TRANSACTION_READ_COMMITTED" />
 
     <Resource name="jdbc/cws"
               auth="Container"
               type="javax.sql.DataSource"
-              factory="org.apache.tomcat.jdbc.pool.DataSourceFactory"
-              uniqueResourceName="process-engine"
+              factory="com.zaxxer.hikari.HikariJNDIFactory"
+              poolName="cws-pool"
+
+              maximumPoolSize="10"
+              connectionTimeout="300000"
+
               driverClassName="__CWS_DB_DRIVER__"
-              url="__CWS_DB_URL__"
+              jdbcUrl="__CWS_DB_URL__"
               username="__CWS_DB_USERNAME__"
               password="__CWS_DB_PASSWORD__"
 
-              initialSize="__CWS_DB_NUM_CONNECTIONS__"
-              minIdle="__CWS_DB_NUM_CONNECTIONS__"
-
-              maxIdle="20"
-              maxActive="20"
-
-              validationQuery="SELECT 1"
-              validationInterval="30000"
-              logValidationErrors="true"
-
-              timeBetweenEvictionRunsMillis="30000"
-              minEvictableIdleTimeMillis="60000"
-              removeAbandoned="true"
-              removeAbandonedTimeout="900"
-
-              testOnBorrow="true"
-              testWhileIdle="true"
-              testOnReturn="true"
-
-              logAbandoned="true"
-
-              defaultTransactionIsolation="READ_COMMITTED" />
+              transactionIsolation="TRANSACTION_READ_COMMITTED" />
 
     <Resource name="global/camunda-bpm-platform/process-engine/ProcessEngineService!org.camunda.bpm.ProcessEngineService" auth="Container"
               type="org.camunda.bpm.ProcessEngineService"

--- a/pom.xml
+++ b/pom.xml
@@ -68,6 +68,7 @@
 		<mockito.version>1.9.5</mockito.version>
 		<mybatis.version>3.5.3</mybatis.version>
 		<mysql-connector.version>8.0.16</mysql-connector.version>
+		<hikaricp.version>4.0.3</hikaricp.version>
 		<phantomjsdriver.version>1.0.4</phantomjsdriver.version>
 		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
 		<quartz.version>2.3.2</quartz.version>
@@ -286,6 +287,19 @@
 				<groupId>mysql</groupId>
 				<artifactId>mysql-connector-java</artifactId>
 				<version>${mysql-connector.version}</version>
+			</dependency>
+
+			<!-- HikariCP database connection pool -->
+			<dependency>
+				<groupId>com.zaxxer</groupId>
+				<artifactId>HikariCP</artifactId>
+				<version>${hikaricp.version}</version>
+				<exclusions>
+					<exclusion>
+						<groupId>org.slf4j</groupId>
+						<artifactId>slf4j-api</artifactId>
+					</exclusion>
+				</exclusions>
 			</dependency>
 
 			<!-- Apache Commons -->

